### ISSUE TITLE
[MONO][PERF] Performance Regression in XML_FromStream benchmarks

### ISSUE THREAD

**Issue by naricc** [dotnet/runtime#56352]
Performance regression in mono-interpreter. As far as I can tell, this is likely caused by https://github.com/dotnet/runtime/commit/ab219299db07b76fc06482bf3a2bfbd7c5f75e63. 

@StephenMolloy does that sound plausible? Is there anything to do about it?

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [b412aac8757d7ac29ab66910e462649eeaecda18](https://github.com/dotnet/runtime/commit/b412aac8757d7ac29ab66910e462649eeaecda18)
Compare | [a8926ba6ecadb259694e09650225d9094081a7df](https://github.com/dotnet/runtime/commit/a8926ba6ecadb259694e09650225d9094081a7df)
Diff | [Diff](https://github.com/dotnet/runtime/compare/b412aac8757d7ac29ab66910e462649eeaecda18...a8926ba6ecadb259694e09650225d9094081a7df)

### Regressions in MicroBenchmarks.Serializers.Xml_FromStream&lt;XmlElement&gt;

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
[XmlSerializer_ - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono/MicroBenchmarks.Serializers.Xml_FromStream(XmlElement).XmlSerializer_.html>) | 47.09 μs | 52.42 μs | 1.11 | 0.00 | True | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(XmlElement)_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(XmlElement).html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'MicroBenchmarks.Serializers.Xml_FromStream&lt;XmlElement&gt;*'
```
<details>

### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-f0154688-116e-4ad8-b09a-7a725410fe769d774381a344c62a0/794565d2-f7ca-46ba-80e9-7dd1e9a985b6.zip?sv=2019-07-07&se=2021-08-13T22%3A24%3A50Z&sr=c&sp=rl&sig=JRDtB8gc4vJQhcuM8qFDnkScEpuKDDGy3NK0obgjXuY%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-49994b26-2651-4fc4-b118-e78ed506c2075f8e50deaf3468bb3/44355bda-a7fa-480a-b87c-800a0671513f.zip?sv=2019-07-07&se=2021-08-14T04%3A41%3A47Z&sr=c&sp=rl&sig=g57q%2FauMYRcar7HvE2Tu1f64ECdLKe79wrHXZRPCDVc%3D>)

#### MicroBenchmarks.Serializers.Xml_FromStream&lt;XmlElement&gt;.XmlSerializer_

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>
### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [b412aac8757d7ac29ab66910e462649eeaecda18](https://github.com/dotnet/runtime/commit/b412aac8757d7ac29ab66910e462649eeaecda18)
Compare | [a8926ba6ecadb259694e09650225d9094081a7df](https://github.com/dotnet/runtime/commit/a8926ba6ecadb259694e09650225d9094081a7df)
Diff | [Diff](https://github.com/dotnet/runtime/compare/b412aac8757d7ac29ab66910e462649eeaecda18...a8926ba6ecadb259694e09650225d9094081a7df)

### Regressions in MicroBenchmarks.Serializers.Xml_FromStream&lt;LoginViewModel&gt;

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
[XmlSerializer_ - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono/MicroBenchmarks.Serializers.Xml_FromStream(LoginViewModel).XmlSerializer_.html>) | 81.46 μs | 94.04 μs | 1.15 | 0.00 | True | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(LoginViewModel)_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(LoginViewModel).html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'MicroBenchmarks.Serializers.Xml_FromStream&lt;LoginViewModel&gt;*'
```
<details>

### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-f0154688-116e-4ad8-b09a-7a725410fe769d774381a344c62a0/794565d2-f7ca-46ba-80e9-7dd1e9a985b6.zip?sv=2019-07-07&se=2021-08-13T22%3A24%3A50Z&sr=c&sp=rl&sig=JRDtB8gc4vJQhcuM8qFDnkScEpuKDDGy3NK0obgjXuY%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-49994b26-2651-4fc4-b118-e78ed506c2075f8e50deaf3468bb3/44355bda-a7fa-480a-b87c-800a0671513f.zip?sv=2019-07-07&se=2021-08-14T04%3A41%3A47Z&sr=c&sp=rl&sig=g57q%2FauMYRcar7HvE2Tu1f64ECdLKe79wrHXZRPCDVc%3D>)

#### MicroBenchmarks.Serializers.Xml_FromStream&lt;LoginViewModel&gt;.XmlSerializer_

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [b412aac8757d7ac29ab66910e462649eeaecda18](https://github.com/dotnet/runtime/commit/b412aac8757d7ac29ab66910e462649eeaecda18)
Compare | [a8926ba6ecadb259694e09650225d9094081a7df](https://github.com/dotnet/runtime/commit/a8926ba6ecadb259694e09650225d9094081a7df)
Diff | [Diff](https://github.com/dotnet/runtime/compare/b412aac8757d7ac29ab66910e462649eeaecda18...a8926ba6ecadb259694e09650225d9094081a7df)

### Regressions in MicroBenchmarks.Serializers.Xml_FromStream&lt;ClassImplementingIXmlSerialiable&gt;

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
[XmlSerializer_ - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono/MicroBenchmarks.Serializers.Xml_FromStream(ClassImplementingIXmlSerialiable).XmlSerializer_.html>) | 44.98 μs | 55.05 μs | 1.22 | 0.00 | True | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(ClassImplementingIXmlSerialiable)_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(ClassImplementingIXmlSerialiable).html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'MicroBenchmarks.Serializers.Xml_FromStream&lt;ClassImplementingIXmlSerialiable&gt;*'
```
<details>

### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-f0154688-116e-4ad8-b09a-7a725410fe769d774381a344c62a0/794565d2-f7ca-46ba-80e9-7dd1e9a985b6.zip?sv=2019-07-07&se=2021-08-13T22%3A24%3A50Z&sr=c&sp=rl&sig=JRDtB8gc4vJQhcuM8qFDnkScEpuKDDGy3NK0obgjXuY%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-49994b26-2651-4fc4-b118-e78ed506c2075f8e50deaf3468bb3/44355bda-a7fa-480a-b87c-800a0671513f.zip?sv=2019-07-07&se=2021-08-14T04%3A41%3A47Z&sr=c&sp=rl&sig=g57q%2FauMYRcar7HvE2Tu1f64ECdLKe79wrHXZRPCDVc%3D>)

#### MicroBenchmarks.Serializers.Xml_FromStream&lt;ClassImplementingIXmlSerialiable&gt;.XmlSerializer_

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [b412aac8757d7ac29ab66910e462649eeaecda18](https://github.com/dotnet/runtime/commit/b412aac8757d7ac29ab66910e462649eeaecda18)
Compare | [a8926ba6ecadb259694e09650225d9094081a7df](https://github.com/dotnet/runtime/commit/a8926ba6ecadb259694e09650225d9094081a7df)
Diff | [Diff](https://github.com/dotnet/runtime/compare/b412aac8757d7ac29ab66910e462649eeaecda18...a8926ba6ecadb259694e09650225d9094081a7df)

### Regressions in MicroBenchmarks.Serializers.Xml_FromStream&lt;Location&gt;

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
[XmlSerializer_ - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono/MicroBenchmarks.Serializers.Xml_FromStream(Location).XmlSerializer_.html>) | 109.43 μs | 118.20 μs | 1.08 | 0.00 | True | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(Location)_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(Location).html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'MicroBenchmarks.Serializers.Xml_FromStream&lt;Location&gt;*'
```
<details>

### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-f0154688-116e-4ad8-b09a-7a725410fe769d774381a344c62a0/794565d2-f7ca-46ba-80e9-7dd1e9a985b6.zip?sv=2019-07-07&se=2021-08-13T22%3A24%3A50Z&sr=c&sp=rl&sig=JRDtB8gc4vJQhcuM8qFDnkScEpuKDDGy3NK0obgjXuY%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-49994b26-2651-4fc4-b118-e78ed506c2075f8e50deaf3468bb3/44355bda-a7fa-480a-b87c-800a0671513f.zip?sv=2019-07-07&se=2021-08-14T04%3A41%3A47Z&sr=c&sp=rl&sig=g57q%2FauMYRcar7HvE2Tu1f64ECdLKe79wrHXZRPCDVc%3D>)

#### MicroBenchmarks.Serializers.Xml_FromStream&lt;Location&gt;.XmlSerializer_

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Baseline | [b412aac8757d7ac29ab66910e462649eeaecda18](https://github.com/dotnet/runtime/commit/b412aac8757d7ac29ab66910e462649eeaecda18)
Compare | [a8926ba6ecadb259694e09650225d9094081a7df](https://github.com/dotnet/runtime/commit/a8926ba6ecadb259694e09650225d9094081a7df)
Diff | [Diff](https://github.com/dotnet/runtime/compare/b412aac8757d7ac29ab66910e462649eeaecda18...a8926ba6ecadb259694e09650225d9094081a7df)

### Regressions in MicroBenchmarks.Serializers.Xml_FromStream&lt;SimpleStructWithProperties&gt;

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
[XmlSerializer_ - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_ubuntu 18.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono/MicroBenchmarks.Serializers.Xml_FromStream(SimpleStructWithProperties).XmlSerializer_.html>) | 78.01 μs | 91.46 μs | 1.17 | 0.00 | True | | |

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(SimpleStructWithProperties)_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_20_2021/refs/heads/main_x64_ubuntu%2018.04_LLVM=false_MonoAOT=false_MonoInterpreter=true_RunKind=micro_mono_Regression/MicroBenchmarks.Serializers.Xml_FromStream(SimpleStructWithProperties).html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
python3 .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'MicroBenchmarks.Serializers.Xml_FromStream&lt;SimpleStructWithProperties&gt;*'
```
<details>

### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-f0154688-116e-4ad8-b09a-7a725410fe769d774381a344c62a0/794565d2-f7ca-46ba-80e9-7dd1e9a985b6.zip?sv=2019-07-07&se=2021-08-13T22%3A24%3A50Z&sr=c&sp=rl&sig=JRDtB8gc4vJQhcuM8qFDnkScEpuKDDGy3NK0obgjXuY%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-49994b26-2651-4fc4-b118-e78ed506c2075f8e50deaf3468bb3/44355bda-a7fa-480a-b87c-800a0671513f.zip?sv=2019-07-07&se=2021-08-14T04%3A41%3A47Z&sr=c&sp=rl&sig=g57q%2FauMYRcar7HvE2Tu1f64ECdLKe79wrHXZRPCDVc%3D>)

#### MicroBenchmarks.Serializers.Xml_FromStream&lt;SimpleStructWithProperties&gt;.XmlSerializer_

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

**Comment by StephenMolloy** [comment]
Seems like the same as #56014

**Comment by SamMonoRT** [comment]
@StephenMolloy - I updated the area label. If same issue, please close one and use the primary one for tracking/fixing.

**Comment by StephenMolloy** [comment]
Should be addressed by #41431.
